tls/extensions/s2n_client_psk.c

```diff
@@ -0,0 +1,250 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <sys/param.h>
+#include <stdint.h>
+
+#include "crypto/s2n_hash.h"
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls_parameters.h"
+
+#include "utils/s2n_safety.h"
+
+#define SIZE_OF_BINDER_SIZE sizeof(uint8_t)
+#define SIZE_OF_BINDER_LIST_SIZE sizeof(uint16_t)
+
+static bool s2n_client_psk_should_send(struct s2n_connection *conn);
+static int s2n_client_psk_send(struct s2n_connection *conn, struct s2n_stuffer *out);
+static int s2n_client_psk_recv(struct s2n_connection *conn, struct s2n_stuffer *extension);
+
+const s2n_extension_type s2n_client_psk_extension = {
+    .iana_value = TLS_EXTENSION_PRE_SHARED_KEY,
+    .is_response = false,
+    .send = s2n_client_psk_send,
+    .recv = s2n_client_psk_recv,
+    .should_send = s2n_client_psk_should_send,
+    .if_missing = s2n_extension_noop_if_missing,
+};
+
+static bool s2n_client_psk_should_send(struct s2n_connection *conn)
+{
+    return conn && s2n_connection_get_protocol_version(conn) >= S2N_TLS13
+            && conn->psk_params.psk_list.len;
+}
+
+static int s2n_client_psk_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+
+    struct s2n_psk_parameters *psk_params = &conn->psk_params;
+    struct s2n_array *psk_list = &psk_params->psk_list;
+
+    struct s2n_stuffer_reservation identity_list_size;
+    GUARD(s2n_stuffer_reserve_uint16(out, &identity_list_size));
+
+    uint16_t binder_list_size = SIZE_OF_BINDER_LIST_SIZE;
+
+    for (size_t i = 0; i < psk_list->len; i++) {
+        struct s2n_psk *psk;
+        GUARD_AS_POSIX(s2n_array_get(psk_list, i, (void**) &psk));
+        notnull_check(psk);
+
+        /* Write the identity */
+        GUARD(s2n_stuffer_write_uint16(out, psk->identity.size));
+        GUARD(s2n_stuffer_write(out, &psk->identity));
+        GUARD(s2n_stuffer_write_uint32(out, psk->obfuscated_ticket_age));
+
+        /* Calculate binder size */
+        uint8_t hash_size;
+        GUARD(s2n_hash_digest_size(psk->hash_alg, &hash_size));
+        binder_list_size += hash_size + SIZE_OF_BINDER_SIZE;
+    }
+
+    GUARD(s2n_stuffer_write_vector_size(&identity_list_size));
+
+    /* Calculating the binders requires a complete ClientHello, and at this point
+     * the extension size, extension list size, and message size are all blank.
+     *
+     * We'll write placeholder data to ensure the extension and extension list sizes
+     * are calculated correctly, then rewrite the binders with real data later. */
+    psk_params->binder_list_size = binder_list_size;
+    GUARD(s2n_stuffer_skip_write(out, binder_list_size));
+
+    return S2N_SUCCESS;
+}
+
+static S2N_RESULT s2n_match_psk_identity(struct s2n_array *known_psks, const struct s2n_blob *wire_identity,
+        struct s2n_psk **match)
+{
+    ENSURE_REF(match);
+    ENSURE_REF(wire_identity);
+    ENSURE_REF(known_psks);
```

Comment:
I'd suggest here
```c
GUARD_RESULT(s2n_array_validate(known_psks));
GUARD_AS_RESULT(s2n_blob_is_valid(wire_identity));
ENSURE_REF(match);
```

Reply:
This method doesn't really care whether `wire_identity` is a valid blob or not, just that it exists. These checks are really just intended to prevent seg faults when we dereference pointers. 

We do care that the blob's data pointer isn't NULL though, as mentioned [here](https://github.com/awslabs/s2n/pull/2409#discussion_r529784710), so I'll add that check.

Reply:
What about `known_psks`? Do we care whether this is a valid structure?
```c
GUARD_RESULT(s2n_array_validate(known_psks));
```